# Attrition analysis

In [88]:
import pandas as pd
import numpy as np

In [89]:
data = pd.read_csv("general_data.csv")
data.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [90]:
data.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

In [103]:
#Checking correlation between the features
data.corr()

Age  Attrition  Department  DistanceFromHome  \
Age                      1.000000  -0.159205   -0.010846          0.006963   
Attrition               -0.159205   1.000000   -0.048206         -0.009730   
Department              -0.010846  -0.048206    1.000000          0.012134   
DistanceFromHome         0.006963  -0.009730    0.012134          1.000000   
Education               -0.035706  -0.015111   -0.004597         -0.008638   
Gender                  -0.039352   0.018125    0.003158         -0.044173   
JobLevel                -0.002884  -0.010290    0.002923         -0.037329   
MaritalStatus           -0.095029   0.162070   -0.044619         -0.027893   
MonthlyIncome           -0.044314  -0.031176   -0.032237         -0.021607   
NumCompaniesWorked       0.299297   0.041503   -0.003958         -0.012617   
PercentSalaryHike       -0.033137   0.032533   -0.007840          0.038125   
StockOptionLevel        -0.031753  -0.006839    0.010512          0.011169   
TotalWorkingYears        0.678463  -0.170744   -0.019026          0.008849   
TrainingTimesLastYear   -0.027308  -0.049431    0.012866         -0.009001   
YearsAtCompany           0.311309  -0.134392    0.010078          0.031684   
YearsSinceLastPromotion  0.216513  -0.033019    0.014845          0.002290   
YearsWithCurrManager     0.202089  -0.156199    0.018016          0.021584   

                         Education    Gender  JobLevel  MaritalStatus  \
Age                      -0.035706 -0.039352 -0.002884      -0.095029   
Attrition                -0.015111  0.018125 -0.010290       0.162070   
Department               -0.004597  0.003158  0.002923      -0.044619   
DistanceFromHome         -0.008638 -0.044173 -0.037329      -0.027893   
Education                 1.000000 -0.016547  0.045746       0.024991   
Gender                   -0.016547  1.000000 -0.026854      -0.009132   
JobLevel                  0.045746 -0.026854  1.000000      -0.022021   
MaritalStatus             0.024991 -0.009132 -0.022021       1.000000   
MonthlyIncome             0.006410  0.007530  0.047316      -0.052383   
NumCompaniesWorked       -0.016096 -0.067927 -0.010877      -0.035065   
PercentSalaryHike        -0.040531  0.011841  0.010973       0.006378   
StockOptionLevel          0.001261  0.019237  0.000993       0.014745   
TotalWorkingYears        -0.011518 -0.027718 -0.035431      -0.076168   
TrainingTimesLastYear     0.010472 -0.030167 -0.032500       0.012798   
YearsAtCompany            0.006080 -0.017277 -0.064219      -0.059986   
YearsSinceLastPromotion   0.022490 -0.021812 -0.060811      -0.030915   
YearsWithCurrManager      0.005358  0.004438 -0.055251      -0.038570   

                         MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  \
Age                          -0.044314            0.299297          -0.033137   
Attrition                    -0.031176            0.041503           0.032533   
Department                   -0.032237           -0.003958          -0.007840   
DistanceFromHome             -0.021607           -0.012617           0.038125   
Education                     0.006410           -0.016096          -0.040531   
Gender                        0.007530           -0.067927           0.011841   
JobLevel                      0.047316           -0.010877           0.010973   
MaritalStatus                -0.052383           -0.035065           0.006378   
MonthlyIncome                 1.000000           -0.019915           0.004325   
NumCompaniesWorked           -0.019915            1.000000           0.031683   
PercentSalaryHike             0.004325            0.031683           1.000000   
StockOptionLevel              0.026930            0.017685           0.012548   
TotalWorkingYears            -0.033635            0.237531          -0.019498   
TrainingTimesLastYear         0.050112           -0.032123          -0.037392   
YearsAtCompany                0.000995           -0.116322          -0.029707   
YearsSinceLast

In [91]:
#droping few unneccesary features 
data.drop(['BusinessTravel', 'EducationField', 'EmployeeID', 'JobRole', 'EmployeeCount', 'Over18', 'StandardHours'], axis=1, inplace=True)

In [92]:
data.head(3) 

Age Attrition              Department  DistanceFromHome  Education  Gender  \
0   51        No                   Sales                 6          2  Female   
1   31       Yes  Research & Development                10          1  Female   
2   32        No  Research & Development                17          4    Male   

   JobLevel MaritalStatus  MonthlyIncome  NumCompaniesWorked  \
0         1       Married         131160                 1.0   
1         1        Single          41890                 0.0   
2         4       Married         193280                 1.0   

   PercentSalaryHike  StockOptionLevel  TotalWorkingYears  \
0                 11                 0                1.0   
1                 23                 1                6.0   
2                 15                 3                5.0   

   TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
0                      6               1                        0   
1                      3               5                        1   
2                      2               5                        0   

   YearsWithCurrManager  
0                     0  
1                     4  
2                     3

In [93]:
data.dtypes #checking dtype 

Age                          int64
Attrition                   object
Department                  object
DistanceFromHome             int64
Education                    int64
Gender                      object
JobLevel                     int64
MaritalStatus               object
MonthlyIncome                int64
NumCompaniesWorked         float64
PercentSalaryHike            int64
StockOptionLevel             int64
TotalWorkingYears          float64
TrainingTimesLastYear        int64
YearsAtCompany               int64
YearsSinceLastPromotion      int64
YearsWithCurrManager         int64
dtype: object

In [94]:
data = data.fillna('NumCompaniesWorked' == 0 )  # Filled null values 

In [95]:
data.describe().T  #Describing the data set 

count          mean           std      min      25%  \
Age                      4410.0     36.923810      9.133301     18.0     30.0   
DistanceFromHome         4410.0      9.192517      8.105026      1.0      2.0   
Education                4410.0      2.912925      1.023933      1.0      2.0   
JobLevel                 4410.0      2.063946      1.106689      1.0      1.0   
MonthlyIncome            4410.0  65029.312925  47068.888559  10090.0  29110.0   
PercentSalaryHike        4410.0     15.209524      3.659108     11.0     12.0   
StockOptionLevel         4410.0      0.793878      0.851883      0.0      0.0   
TrainingTimesLastYear    4410.0      2.799320      1.288978      0.0      2.0   
YearsAtCompany           4410.0      7.008163      6.125135      0.0      3.0   
YearsSinceLastPromotion  4410.0      2.187755      3.221699      0.0      0.0   
YearsWithCurrManager     4410.0      4.123129      3.567327      0.0      2.0   

                             50%      75%       max  
Age                         36.0     43.0      60.0  
DistanceFromHome             7.0     14.0      29.0  
Education                    3.0      4.0       5.0  
JobLevel                     2.0      3.0       5.0  
MonthlyIncome            49190.0  83800.0  199990.0  
PercentSalaryHike           14.0     18.0      25.0  
StockOptionLevel             1.0      1.0       3.0  
TrainingTimesLastYear        3.0      3.0       6.0  
YearsAtCompany               5.0      9.0      40.0  
YearsSinceLastPromotion      1.0      3.0      15.0  
YearsWithCurrManager         3.0      7.0      17.0

# will convert the object dtypes to int dtypes

In [48]:
from sklearn import preprocessing

In [96]:
label_enoder = preprocessing.LabelEncoder()
data['Attrition'] = label_enoder.fit_transform(data['Attrition'])  #Converted Attrition to int type from object type
data['Department'] = label_enoder.fit_transform(data['Department']) #converted department to int type as(1,2,3)
data['Gender'] = label_enoder.fit_transform(data['Gender']) #converted gender to int type 
data['MaritalStatus'] = label_enoder.fit_transform(data['MaritalStatus']) #converted marital status to int type
data['NumCompaniesWorked'] = label_enoder.fit_transform(data['NumCompaniesWorked'])
data['TotalWorkingYears'] = label_enoder.fit_transform(data['TotalWorkingYears'])

In [97]:
data.head(8)

Age  Attrition  Department  DistanceFromHome  Education  Gender  JobLevel  \
0   51          0           2                 6          2       0         1   
1   31          1           1                10          1       0         1   
2   32          0           1                17          4       1         4   
3   38          0           1                 2          5       1         3   
4   32          0           1                10          1       1         1   
5   46          0           1                 8          3       0         4   
6   28          1           1                11          2       1         2   
7   29          0           1                18          3       1         2   

   MaritalStatus  MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  \
0              1         131160                   1                 11   
1              2          41890                   0                 23   
2              1         193280                   1                 15   
3              1          83210                   3                 11   
4              2          23420                   4                 12   
5              1          40710                   3                 13   
6              2          58130                   2                 20   
7              1          31430                   2                 22   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0                  1                      6               1   
1                 1                  6                      3               5   
2                 3                  5                      2               5   
3                 3                 13                      5               8   
4                 2                  9                      2               6   
5                 0                 28                      5               7   
6                 1                  5                      2               0   
7                 3                 10                      2               0   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  
5                        7                     7  
6                        0                     0  
7                        0                     0

# Now we will start the Logistic Regression Anylysis

## Assigning DV & IDV

In [99]:
y  = data['Attrition']  #dependent variable
x = data[['Age', 'Department', 'DistanceFromHome', 'Education', 'Gender', 'JobLevel', 'MaritalStatus', 'MonthlyIncome', 'NumCompaniesWorked', 'PercentSalaryHike', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager']] #independent variable

In [100]:
#Adding constant 
import statsmodels.api as sm
x1 = sm.add_constant(x)

In [101]:
Logistic = sm.Logit(y,x1)

In [102]:
Result = Logistic.fit()

Optimization terminated successfully.
         Current function value: 0.394427
         Iterations 7


In [104]:
Result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Attrition   No. Observations:                 4410
Model:                          Logit   Df Residuals:                     4393
Method:                           MLE   Df Model:                           16
Date:                Sun, 09 Aug 2020   Pseudo R-squ.:                  0.1070
Time:                        21:51:19   Log-Likelihood:                -1739.4
converged:                       True   LL-Null:                       -1947.9
Covariance Type:            nonrobust   LLR p-value:                 1.039e-78
===========================================================================================
                              coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------
const                      -0.0006      0.377     -0.002      0.999      -0.740       0.739
Age                        -0.0310      0.007     -4.571      0.000      -0.044      -0.018
Department                 -0.2636      0.081     -3.241      0.001      -0.423      -0.104
DistanceFromHome           -0.0012      0.005     -0.214      0.831      -0.012       0.009
Education                  -0.0585      0.042     -1.378      0.168      -0.142       0.025
Gender                      0.0915      0.089      1.025      0.305      -0.083       0.267
JobLevel                   -0.0223      0.040     -0.563      0.573      -0.100       0.055
MaritalStatus               0.5988      0.063      9.512      0.000       0.475       0.722
MonthlyIncome           -1.822e-06   9.53e-07     -1.912      0.056   -3.69e-06    4.61e-08
NumCompaniesWorked          0.1155      0.018      6.327      0.000       0.080       0.151
PercentSalaryHike           0.0114      0.012      0.976      0.329      -0.012       0.034
StockOptionLevel           -0.0634      0.051     -1.234      0.217      -0.164       0.037
TotalWorkingYears          -0.0583      0.012     -4.978      0.000      -0.081      -0.035
TrainingTimesLastYear      -0.1403      0.035     -4.016      0.000      -0.209      -0.072
YearsAtCompany              0.0118      0.018      0.645      0.519      -0.024       0.047
YearsSinceLastPromotion     0.1286      0.020      6.307      0.000       0.089       0.169
YearsWithCurrManager       -0.1378      0.022     -6.241      0.000      -0.181      -0.095
===========================================================================================
"""

# Significance based on p value are -:
1. Age = It's p-value == 0, it has Heigher Significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$ (p-value = 0.000)   
2. Department = It's p-value < 0.05, it has consederable/much significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.001)
3. DistanceFromHome = It's p-value > 0.05, it has not much significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.831)
4. Education =  It's p-value > 0.05, it has not much significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.168)
5. Gender =  It's p-value > 0.05, it has not much significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.305)
6. JobLevel =  It's p-value > 0.05, it has not much significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.573)
7. MaritalStatus = It's p-value == 0, it has Heigher Significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.000)
8. MonthlyIncome = It's p-value > 0.05, it has not much significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.056)
9. NumCompaniesWorked = It's p-value == 0, it has Heigher Significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.000)
10. PercentSalaryHike = It's p-value > 0.05, it has not much significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.329)
11. StockOptionLevel = It's p-value > 0.05, it has not much significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.217)
12. TotalWorkingYears =  It's p-value == 0, it has Heigher Significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.000)
13. TrainingTimesLastYear = It's p-value == 0, it has Heigher Significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.000)
14. YearsAtCompany	= It's p-value > 0.05, it has not much significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.519)
15. YearsSinceLastPromotion = It's p-value == 0, it has Heigher Significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.000)
16. YearsWithCurrManager = It's p-value == 0, it has Heigher Significance.$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$(p-value = 0.000)